File: server/src/main/java/org/codi/catan/model/game/BaseState.java
/*
 * @author the-codinator
 * created on 2020/5/26
 */

package org.codi.catan.model.game;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import java.util.EnumMap;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.codi.catan.model.core.IdentifiableEntity;
import org.codi.catan.model.core.StrongEntity;

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
public class BaseState implements IdentifiableEntity, StrongEntity {

    public BaseState(String id) {
        this.id = id;
    }

    private String id;
    private Phase phase;
    private List<House> houses;
    private List<Road> roads;
    private int thief;
    @JsonInclude(Include.NON_NULL)
    private EnumMap<Resource, Integer> bank;
    @JsonInclude(Include.NON_NULL)
    private EnumMap<Color, List<DevCard>> playedDevCards;
    @JsonInclude(Include.NON_NULL)
    private EnumMap<AchievementType, AchievementValue> achievements;
    private CurrentMove currentMove;
    @SuppressWarnings("checkstyle:MemberName")
    private String eTag;

    @Override
    @JsonIgnore
    public String getETag() {
        return this.eTag;
    }
}

File: server-azure/src/model/game/base-state.ts
import type { AchievementType, AchievementValue } from './achievement';
import type { IdentifiableEntity, StrongEntity } from '../core';
import type { Color } from './color';
import type { CurrentMove } from './current-move';
import type { DevCard } from './dev-card';
import type { House } from './house';
import type { Phase } from './phase';
import type { Resource } from './resource';
import type { Road } from './road';

export interface BaseState extends IdentifiableEntity, StrongEntity {
  phase: Phase;
  houses: Partial<Record<number, House>>;
  roads: Road[];
  thief: number;
  bank: Record<Resource, number>;
  playedDevCards: Partial<Record<Color, DevCard[]>>;
  achievements: Record<AchievementType, AchievementValue>;
  currentMove: CurrentMove;
}
